test/e2e/node_drain_timeout.go

```diff
@@ -0,0 +1,200 @@
+				Flavor:                   clusterctl.DefaultFlavor,
+				Namespace:                namespace.Name,
+				ClusterName:              fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersion),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(1),
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+		cluster = applyClusterTemplateResult.Cluster
+		controlplane = applyClusterTemplateResult.ControlPlane
+		machineDeployments = applyClusterTemplateResult.MachineDeployments
+
+		By("Update the nodeDrainTimeout field of the machinedeployment and wait for all machines to be updated")
+
+		nodeDrainTimeoutSecond := 60
+		nodeDrainTimeoutDuration := durationMaker(durationMakerInput{
+			second: float64(nodeDrainTimeoutSecond),
+		})
+		nodeDrainTimeoutMachineDeploymentInterval := convertMachineDeploymentDurationToInterval(nodeDrainTimeoutDuration)
+
+		framework.UpdateNodeDrainTimeoutInMachineDeployment(ctx, framework.UpdateNodeDrainTimeoutInMachineDeploymentInput{
+			ClusterProxy:               input.BootstrapClusterProxy,
+			Cluster:                    cluster,
+			MachineDeployments:         machineDeployments,
+			NodeDrainTimeout:           nodeDrainTimeoutDuration,
+			WaitForMachinesToBeUpdated: input.E2EConfig.GetIntervals(specName, "wait-machine-updated"),
+		})
+
+		By("Add a deployment and podDisruptionBudget to the workload cluster. The deployed pods cannot be evicted in the node draining process.")
+		framework.AddUnevictablePod(ctx, framework.AddUnevictablePodInput{
+			ClusterProxy:                       input.BootstrapClusterProxy,
+			Cluster:                            cluster,
+			MachineDeployments:                 machineDeployments,
+			WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
+		})
+
+		By("Scale the machinedeployment down to zero. If we didn't have the NodeDrainTimeout duration, the node drain process would block this operator.")
+		for _, md := range machineDeployments {
+			framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+				ClusterProxy:              input.BootstrapClusterProxy,
+				Cluster:                   cluster,
+				MachineDeployment:         md,
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Update nodeDrainTimeout field of the existing and new controlplane machines.")
+		numScaledUpControlPlane := 3
+		framework.UpdateControlplaneNodeDrainTimeout(ctx, framework.UpdateControlplaneNodeDrainTimeoutInput{
```

Comment:
Node drain timeout for control plane and machine deployment can be embedded into cluster-template node-drain flavor so that we do not have to patch them.